### DISCUSSION TITLE
add an option to follow soft links

### DISCUSSION THREAD

**Discussion by wagde-orca** [aquasecurity/trivy#4505] [Ideas]
### Description

add an option in rootfs/fs to allow following soft links... this will help in scanning files that are found outside the root and have a link inside the root...
it does not work in the latest version and it is easy to reproduce

scanning a dir with a jar finds the file and scan it
trivy   rootfs --vuln-type library --list-all-pkgs  -o res.json -f json --scanners vuln  /tmp/jar_file
2023-05-30T19:08:27.364+0300	INFO	Vulnerability scanning is enabled
2023-05-30T19:08:27.365+0300	INFO	JAR files found
2023-05-30T19:08:27.365+0300	INFO	Analyzing JAR files takes a while...
2023-05-30T19:08:27.771+0300	INFO	Number of language-specific files: 1
2023-05-30T19:08:27.771+0300	INFO	Detecting jar vulnerabilities...

scanning a dir with link does not find the file
ls -ltr /tmp/jar_link
lrwxr-xr-x  1 wagdezabit  wheel      19 May 30 19:07 a.jar -> /tmp/jar_file/a.jar
trivy   rootfs --vuln-type library --list-all-pkgs  -o res.json -f json --scanners vuln  /tmp/jar_link
2023-05-30T19:09:48.244+0300	INFO	Vulnerability scanning is enabled
2023-05-30T19:09:48.252+0300	INFO	Number of language-specific files: 0

### Target

Filesystem

### Scanner

Vulnerability

**Comment by knqyf263** [comment]
Following symlinks leads to duplicated results if the actual entity is in the scanning directory.

**Comment by wagde-orca** [comment]
@knqyf263 you are right, but maybe we can add a verification that the actual entry is not in the results and add it
i think this is a good idea not to have duplicates in the results...
for example if you have the same jar file in two directories we will have two entries in Packages and two entries in Vulnerabilities..
instead we can have one entry with multiple FilePaths and not only one...

See this example
find .
./a
./a/jackson-databind-2.9.8.jar
./b
./b/jackson-databind-2.9.8.jar

trivy   rootfs --vuln-type library --list-all-pkgs  -o res.json -f json --scanners vuln .

we will have
      "Packages": [
        {
          "Name": "com.fasterxml.jackson.core:jackson-databind",
          "Version": "2.9.8",
          "Layer": {},
          "FilePath": "a/jackson-databind-2.9.8.jar"
        },
        {
          "Name": "com.fasterxml.jackson.core:jackson-databind",
          "Version": "2.9.8",
          "Layer": {},
          "FilePath": "b/jackson-databind-2.9.8.jar"
        }
      ],

this can lead to huge result json in case the same jar file exist in tens/hundreds of directories...
I hope you will consider this optimization 

anyway... I suggest adding following symlinks as an option and not the default behavior :-)

**Comment by itaysk** [comment]
also possibly addresses this concern: 

https://github.com/bgeesaman/malicious-compliance/blob/main/demo/demo3.gif
https://github.com/bgeesaman/malicious-compliance/blob/main/docker/Dockerfile-3-lang

**Comment by OdinSongeThorsen** [comment]
Following up on this — I'd like to propose a concrete, narrow version, and I've opened a PR with a working implementation + measurements.

**Motivation (least privilege, not just convenience).** A common need is scanning a filesystem *assembled* from content that lives elsewhere — e.g. a container image's effective rootfs reconstructed from a content-addressed store (Nix is the obvious case). Today that means either a full **copy** (doubles disk) or a **`mount --bind`**, which needs `CAP_SYS_ADMIN` — real root or an *unprivileged user namespace*. The cheap, capability-free option — a **symlink** layout — doesn't work, because `fs`/`rootfs` skip symlinks (`filepath.WalkDir` lstat's and won't follow them). And unprivileged userns is increasingly **restricted by default**: Ubuntu 23.10+ AppArmor restrictions, `user.max_user_namespaces=0` / `kernel.unprivileged_userns_clone=0` hardening, container seccomp blocking `CLONE_NEWUSER`, hardened kernels. Following symlinks lets a scan root be assembled with **plain file operations — no `mount`, no `CAP_SYS_ADMIN`, no namespace.**

**Proposal.** An **opt-in** `--follow-symlinks` (default **off**) that follows symlinks **resolving to regular files only**. Symlinks to **directories are not followed** — that's where cycles come from, so skipping them sidesteps the loop concern; default behavior is unchanged. This also addresses the duplicate-results concern raised earlier: it's opt-in, and in the assembled-root case the targets live *outside* the scan root, so there's no in-tree double-count.

**Evidence (prototyped).** On a real Nix-built image rootfs laid out as a symlink farm: stock Trivy detects **0 packages** (symlinks skipped); with `--follow-symlinks`, detection is **identical to both a full dereferenced copy and a `mount --bind` assembly** — same package set, package-for-package — at a fraction of the disk and with no `mount`/userns requirement.

Implementation is small (a walker option + a scan flag). **PR: #10743**
